### PR TITLE
Improve user/group creation in image build

### PR DESCRIPTION
**Pull request by ellie-di**
This PR is a combination of the following changes that's been in review process once:
1- https://github.com/microsoft/azurelinux/pull/6957
2- https://github.com/microsoft/azurelinux/pull/6876

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
This PR accomplishes the following goals:
1- Add support to specify home directory and primary group when creating users. We needed this to overwrite those defaults for some of the service's users that are predefined in an Azure Linux derivative image. Also, added a validation step in imageconfigvalidator to ensure "shadow-utils" package is added to the packages list. 
2- Add support to define a list of users and groups when building a rootFS image type and do it earlier in the image build stage. The reason for shifting users/groups creation to an earlier time is that some of the upstream packages also define their own users or groups, and the chances are some of the users/groups get defined by installing packages in the build chroot and later on addUsers/addGroups will fail because some of the users/groups are already defined. Also, added a validation step in imageconfigvalidator to enforce each user can only belong to the groups that are defined in the config.

###### Change Log  

- Change
- Change
- Change

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

Tested locally to build an image with a list of users with predefined home directory and primary group

### CODE REVIEW

**Review by snambakam — APPROVED** [review]

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 1]
```diff
@@ -493,6 +494,15 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
 		return
 	}
 
+	if len(config.Users) > 0 || len(config.Groups) > 0 {
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/installutils/installutils.go:497]: Maybe something along these lines? Or failing that ensure that `shadow-utils` is in the install list in code.

```suggestion
	// imageconfigvalidator should have ensured that we intend to install shadow-utils, so we ca go ahead and do that here.
	if len(config.Users) > 0 || len(config.Groups) > 0 {
```

Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:497]: Or
```suggestion
	if len(config.Users) > 0 || len(config.Groups) > 0 {
		if ! sliceutils.ContainsValue(packagesToInstall, "shadow-utils") {
			err = fmt.Errorf("shadow-utils package is required when setting users or groups")
			return
		}
```

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 2]
```diff
@@ -493,6 +494,15 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
...
+		shadowUtilsInstalled := 0
+		shadowUtilsInstalled, err = TdnfInstallWithProgress(shadowUtilsPkg, installRoot, packagesInstalled, totalPackages, true)
+		if err != nil {
+			return
+		}
+		packagesInstalled += shadowUtilsInstalled
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/installutils/installutils.go:503]: Looks like `TdnfInstallWithProgress()` takes care of incrementing the number of packages if you pass the current number in.
```suggestion
		packagesInstalled, err = TdnfInstallWithProgress(shadowUtilsPkg, installRoot, packagesInstalled, totalPackages, true)
		if err != nil {
			return
		}
```

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 3]
```diff
@@ -493,6 +494,15 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
...
+	}
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/installutils/installutils.go:504]: Bit of a nit, and we don't do this for the filesystem above so... 🤷
Also not 100% sure this works (I think its good though...)
```suggestion
		// Remove shadow-utils from the list of packages to install
		packagesToInstall = sliceutils.FindMatches(packagesToInstall, func(pkg string) bool { return pkg != "shadow-utils" })

	}
```

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 1 above)
Comment (ellie-di) [reply toolkit/tools/imagegen/installutils/installutils.go:497]: Done.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 3 above)
Comment (ellie-di) [reply toolkit/tools/imagegen/installutils/installutils.go:512]: Applied this suggestion anyway and confirmed it works.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 4]
```diff
@@ -493,6 +495,22 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
...
+			return
+		}
+
+		packagesInstalled, err = TdnfInstallWithProgress(shadowUtilsPkg, installRoot, packagesInstalled, totalPackages, true)
+		if err != nil {
+			return
+		}
+
+		// Remove shadow-utils from the list of packages to install
+		packagesToInstall = sliceutils.FindMatches(packagesToInstall, func(pkg string) bool { return pkg != "shadow-utils" })
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/installutils/installutils.go:511]: Sorry, forgot there was a new const. We should use that.
```suggestion
		packagesToInstall = sliceutils.FindMatches(packagesToInstall, func(pkg string) bool { return pkg != shadowUtilsPkg })
```

**Review by jslobodzian — APPROVED** [review]

File: toolkit/tools/internal/userutils/userutils.go [hunk 5]
```diff
@@ -76,14 +76,20 @@ func UserExists(username string, installChroot safechroot.ChrootInterface) (bool
 	return userExists, nil
 }
 
-func AddUser(username string, hashedPassword string, uid string, installChroot safechroot.ChrootInterface) error {
+func AddUser(username string, homeDir string, primaryGroup string, hashedPassword string, uid string, installChroot *safechroot.Chroot) error {
```
Comment (cwize1) [inline toolkit/tools/internal/userutils/userutils.go:79]: Why did you change `installChroot` from `ChrootInterface` to `*safechroot.Chroot`? Please revert.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 6]
```diff
@@ -502,6 +520,18 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
 		}
 	}
 
+	// Add groups
+	err = addGroups(installChroot, config.Groups)
```
Comment (cwize1) [inline toolkit/tools/imagegen/installutils/installutils.go:524]: What it intentional to move `addGroups` out of the `if !config.IsRootFS()` check?

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 7]
```diff
@@ -1380,6 +1398,9 @@ func addGroups(installChroot *safechroot.Chroot, groups []configuration.Group) (
 		err = installChroot.UnsafeRun(func() error {
 			return shell.ExecuteLive(squashErrors, "groupadd", args...)
 		})
+		if err != nil {
+			return fmt.Errorf("failed to add group: %w", err)
```
Comment (cwize1) [inline toolkit/tools/imagegen/installutils/installutils.go:1402]: nit: `"failed to add group:\n%w"`

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 8]
```diff
@@ -1476,7 +1497,7 @@ func createUserWithPassword(installChroot *safechroot.Chroot, user configuration
 		}
 		isRoot = true
 	} else {
-		err = userutils.AddUser(user.Name, hashedPassword, user.UID, installChroot)
+		err = userutils.AddUser(user.Name, user.HomeDirectory, user.PrimaryGroup, hashedPassword, user.UID, installChroot)
```
Comment (cwize1) [inline toolkit/tools/imagegen/installutils/installutils.go:1500]: If you are going to set the user's primary group here, then you should skip the `ConfigureUserGroupMembership` call in `addUsers` when the user is added (as opposed to the user being updated).

File: toolkit/tools/imagegen/configuration/systemconfig.go [hunk 9]
```diff
@@ -66,6 +67,34 @@ func (s *SystemConfig) GetMountpointPartitionSetting(mountPoint string) (partiti
...
+				logger.Log.Warnf("Primary group (%s) for user (%s) not defined", user.PrimaryGroup, user.Name)
+			}
+		}
+
+		for _, group := range user.SecondaryGroups {
+			if !sliceutils.Contains(s.Groups, group, groupMatchFunc) {
+				// Unclear how to validate these while allowing users to be added to existing system groups. If we
+				// define a group in the config that already exists it will cause errors.
+				// Maybe we can scrape /etc/group and /etc/passwd from filesystem.sepc to validate these?
+				logger.Log.Warnf("secondary group (%s) for user (%s) not defined", group, user.Name)
```
Comment (cwize1) [inline toolkit/tools/imagegen/configuration/systemconfig.go:90]: nit: `Secondary`. (Logs start with an uppercase letter. Errors start with a lowercase letter.)

File: toolkit/tools/imagegen/configuration/systemconfig.go [hunk 10]
```diff
@@ -66,6 +67,34 @@ func (s *SystemConfig) GetMountpointPartitionSetting(mountPoint string) (partiti
...
+		groupMatchFunc := func(groupName interface{}, groupObj interface{}) bool {
+			return groupName == groupObj.(Group).Name
+		}
+
+		if user.PrimaryGroup != "" {
+			if !sliceutils.Contains(s.Groups, user.PrimaryGroup, groupMatchFunc) {
+				// Unclear how to validate these while allowing users to be added to existing system groups. If we
+				// define a group in the config that already exists it will cause errors.
+				// Maybe we can scrape /etc/group and /etc/passwd from filesystem.sepc to validate these?
+				logger.Log.Warnf("Primary group (%s) for user (%s) not defined", user.PrimaryGroup, user.Name)
```
Comment (cwize1) [inline toolkit/tools/imagegen/configuration/systemconfig.go:81]: "not defined" is probably specific enough. Perhaps, "not defined in SystemConfig.Groups"?

File: toolkit/tools/pkg/imagecustomizerlib/customizeutils.go [hunk 11]
```diff
@@ -315,7 +315,7 @@ func addOrUpdateUser(user imagecustomizerapi.User, baseConfigPath string, imageC
 		}
 
 		// Add the user.
-		err = userutils.AddUser(user.Name, hashedPassword, uidStr, imageChroot)
+		err = userutils.AddUser(user.Name, user.HomeDirectory, user.PrimaryGroup, hashedPassword, uidStr, imageChroot.(*safechroot.Chroot))
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeutils.go:318]: You need to handle the home directory for the case where the user is updated instead of added.

If it isn't possible (or practical) to move the home directory, then you should report an error if the home directory value is specified.

Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeutils.go:318]: The `ConfigureUserGroupMembership` call below should be skipped if the user is added.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/imageconfigvalidator/imageconfigvalidator.go [hunk 12]
```diff
@@ -170,6 +175,11 @@ func validatePackages(config configuration.Config) (err error) {
 				return fmt.Errorf("%s: [SELinux] selected, but '%s' package is not included in the package lists", validateError, selinuxPkgName)
 			}
 		}
+		if len(systemConfig.Users) > 0 || len(systemConfig.Groups) > 0 {
+			if !foundUserAddPackage {
+				return fmt.Errorf("%s: add users require '%s' package that is not included in the package lists", validateError, userAddPkgName)
```
Comment (PawelWMS) [inline toolkit/tools/imageconfigvalidator/imageconfigvalidator.go:180]: Suggestion to make it clearer for the user what's missing:
```suggestion
				return fmt.Errorf("%s: the '%s' package must be included in the package lists when the image is configured to add users or groups", validateError, userAddPkgName)
```

File: toolkit/tools/imageconfigvalidator/imageconfigvalidator_test.go [hunk 13]
```diff
@@ -301,3 +301,44 @@ func TestShouldSucceedSELinuxPackageDefinedInline(t *testing.T) {
...
+
+			err = ValidateConfiguration(config)
+			assert.Error(t, err)
+			assert.Equal(t, "failed to validate package lists in config: add users require 'shadow-utils' package that is not included in the package lists", err.Error())
+
+			return
+		}
+	}
+	assert.Fail(t, "Could not find "+targetPackage+" to test")
+}
```
Comment (PawelWMS) [inline toolkit/tools/imageconfigvalidator/imageconfigvalidator_test.go:344]: Why use non-test files and modify them by code here? Why not just use a test config file, which will be dedicated just for the tests? We're adding unnecessary logic in the test code and mixing test and non-test components.

File: toolkit/tools/imagegen/configuration/systemconfig.go [hunk 14]
```diff
@@ -66,6 +67,34 @@ func (s *SystemConfig) GetMountpointPartitionSetting(mountPoint string) (partiti
 	return FindMountpointPartitionSetting(s.PartitionSettings, mountPoint)
 }
 
+func (s *SystemConfig) validateUsersAndGroups() (err error) {
+	for _, user := range s.Users {
+		groupMatchFunc := func(groupName interface{}, groupObj interface{}) bool {
+			return groupName == groupObj.(Group).Name
+		}
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/systemconfig.go:74]: Is there a need to define `groupMatchFunc` for each iteration of the loop?
```suggestion
	groupMatchFunc := func(groupName interface{}, groupObj interface{}) bool {
		return groupName == groupObj.(Group).Name
	}

	for _, user := range s.Users {
```

File: toolkit/tools/imagegen/configuration/systemconfig.go [hunk 15]
```diff
@@ -66,6 +67,34 @@ func (s *SystemConfig) GetMountpointPartitionSetting(mountPoint string) (partiti
...
+
+		if user.PrimaryGroup != "" {
+			if !sliceutils.Contains(s.Groups, user.PrimaryGroup, groupMatchFunc) {
+				// Unclear how to validate these while allowing users to be added to existing system groups. If we
+				// define a group in the config that already exists it will cause errors.
+				// Maybe we can scrape /etc/group and /etc/passwd from filesystem.sepc to validate these?
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/systemconfig.go:80]: Can you be a bit more clear with the comment? What do you mean by "these" in "Unclear how to validate these" for instance? Also, I'm not sure, if the last comments is helpful. If we can check the files, let's do it. If not, why add the comment?

File: toolkit/tools/imagegen/configuration/systemconfig.go (hunk 9 above)
Comment (PawelWMS) [reply toolkit/tools/imagegen/configuration/systemconfig.go:90]: Also add a period at the end, please.

File: toolkit/tools/imagegen/configuration/systemconfig.go (hunk 10 above)
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/systemconfig.go:81]: I'd also add a bit of explanation to the warning saying what is expected and what will happen, if the expectation is not met.

File: toolkit/tools/imagegen/configuration/user.go [hunk 16]
```diff
@@ -25,6 +25,7 @@ type User struct {
 	PrimaryGroup        string   `json:"PrimaryGroup"`
 	SecondaryGroups     []string `json:"SecondaryGroups"`
 	StartupCommand      string   `json:"StartupCommand"`
+	HomeDirectory       string   `json:"HomeDirectory"`
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/user.go:28]: Is there documentation we should extend for this new field?

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 17]
```diff
@@ -493,6 +495,22 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
 		return
 	}
 
+	// imageconfigvalidator should have ensured that we intend to install shadow-utils, so we can go ahead and do that here.
+	if len(config.Users) > 0 || len(config.Groups) > 0 {
+		if !sliceutils.ContainsValue(packagesToInstall, "shadow-utils") {
+			err = fmt.Errorf("shadow-utils package is required when setting users or groups")
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:501]: I'd mention that `shado-utils` must be listed explicitly. Otherwise, it's going to be confusing, if someone adds a package, which pulls in `shadow-utils`.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 18]
```diff
@@ -493,6 +495,22 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
...
+		}
+
+		packagesInstalled, err = TdnfInstallWithProgress(shadowUtilsPkg, installRoot, packagesInstalled, totalPackages, true)
+		if err != nil {
+			return
+		}
+
+		// Remove shadow-utils from the list of packages to install
+		packagesToInstall = sliceutils.FindMatches(packagesToInstall, func(pkg string) bool { return pkg != shadowUtilsPkg })
+	}
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:512]: This section looks long and complex enough to move to a separate function.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 19]
```diff
@@ -493,6 +495,22 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
 		return
 	}
 
+	// imageconfigvalidator should have ensured that we intend to install shadow-utils, so we can go ahead and do that here.
+	if len(config.Users) > 0 || len(config.Groups) > 0 {
+		if !sliceutils.ContainsValue(packagesToInstall, "shadow-utils") {
+			err = fmt.Errorf("shadow-utils package is required when setting users or groups")
+			return
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:502]: Doesn't this belong in the config validation code earlier on?

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 20]
```diff
@@ -493,6 +495,22 @@ func PopulateInstallRoot(installChroot *safechroot.Chroot, packagesToInstall []s
...
+		}
+
+		packagesInstalled, err = TdnfInstallWithProgress(shadowUtilsPkg, installRoot, packagesInstalled, totalPackages, true)
+		if err != nil {
+			return
+		}
+
+		// Remove shadow-utils from the list of packages to install
+		packagesToInstall = sliceutils.FindMatches(packagesToInstall, func(pkg string) bool { return pkg != shadowUtilsPkg })
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/installutils/installutils.go:511]: Why do we need to install `shadow-utils` separately? And why do we remove it later from `packagesToInstall`?

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 7 above)
Comment (PawelWMS) [reply toolkit/tools/imagegen/installutils/installutils.go:1402]: Please also say which group we failed to add. It's not guaranteed that name is going to be mentioned the output from `groupadd`.

File: toolkit/tools/imageconfigvalidator/imageconfigvalidator_test.go (hunk 13 above)
Comment (dmcilvaney) [reply toolkit/tools/imageconfigvalidator/imageconfigvalidator_test.go:344]: Don't disagree, but I don't want to re-invent how we do testing for the validator right now. This is copy/pasted from another test above.

File: toolkit/tools/imagegen/configuration/systemconfig.go (hunk 14 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/configuration/systemconfig.go:74]: Readability is better to me, and this is such a small loop the perf uplift didn't seem important. I originally had it as an inline function in the call to sliceutils but it was a bit hard to read.

File: toolkit/tools/imagegen/configuration/systemconfig.go (hunk 15 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/configuration/systemconfig.go:80]:
```suggestion
				// Unclear how to validate that users and groups match together correctly while still allowing
				// users to be added to existing system groups. If we define a group in the config that already
				// exists it will cause errors, but we don't have insight into the default users/groups already in
				// an image before we install the filesystem package.
```

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 17 above)
Comment (dmcilvaney) [inline toolkit/tools/imagegen/installutils/installutils.go:501]:
```suggestion
			err = fmt.Errorf("shadow-utils package must be added to the image's package lists when setting users or groups")
```

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 19 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:502]: We put as much validation as possible natively as part of each config structure. The validator tool will use these functions when it runs its validation, but the config validator's job is to check against external things that aren't encoded in the .json. All the user bits are part of the .json. This is a bit odd since its now just a warning while it was originally an actual error.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 20 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:511]: The point of this change is to add users before installing any other packages. To do that we need shadow-utils (so at best we can try to minimize the number of other packages added). It feels cleaner to not re-install shadowutils if its already installed.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 18 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:512]: Can fix later, we need to get this merged for testing as early as possible.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 7 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:1402]:
```suggestion
			return fmt.Errorf("failed to add group (%s):\n%w",  group.Name, err)
```

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 6 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:524]: Unclear why we would want to gate groups behind rootfs while we don't with users.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 8 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:1500]: Removed and renamed function to `ConfigureUserSecondaryGroupMembership()`.

File: toolkit/tools/internal/userutils/userutils.go (hunk 5 above)
Comment (dmcilvaney) [reply toolkit/tools/internal/userutils/userutils.go:79]: Sorry, didn't see that when I was going through the changes. Will revert.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 8 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:1500]: Actually... that might break MIC...

File: toolkit/tools/pkg/imagecustomizerlib/customizeutils.go (hunk 11 above)
Comment (dmcilvaney) [reply toolkit/tools/pkg/imagecustomizerlib/customizeutils.go:318]: We need to configure secondary groups still, even if the primary group is ok. I'm wondering if the extra code complexity is worth it here to decide if we need to pass "" or `user.PrimaryGroup` in...

Comment (dmcilvaney) [reply toolkit/tools/pkg/imagecustomizerlib/customizeutils.go:318]: ok, added conditional `var primaryGroup string`.

File: toolkit/tools/imagegen/installutils/installutils.go (hunk 8 above)
Comment (dmcilvaney) [reply toolkit/tools/imagegen/installutils/installutils.go:1500]: Split into `ConfigureUserPrimaryGroupMembership()` and `ConfigureUserSecondaryGroupMembership()` so we can target what we need.

File: toolkit/tools/imagecustomizerapi/user.go [hunk 21]
```diff
@@ -19,6 +19,7 @@ type User struct {
 	PrimaryGroup        string    `yaml:"primaryGroup"`
 	SecondaryGroups     []string  `yaml:"secondaryGroups"`
 	StartupCommand      string    `yaml:"startupCommand"`
+	HomeDirectory       string    `yaml:"HomeDirectory"`
```
Comment (cwize1) [inline toolkit/tools/imagecustomizerapi/user.go:22]: `homeDirectory`

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/imagecustomizerapi/user.go (hunk 21 above)
Comment (dmcilvaney) [reply toolkit/tools/imagecustomizerapi/user.go:22]:
```suggestion
	HomeDirectory       string    `yaml:"homeDirectory"`
```

**Review by cwize1 — APPROVED** [review]
